rAF callback arg vs performance.now(), fixes my previous mistake (#28590)
mdn/content: files/en-us/web/api/window/requestanimationframe/index.md

{{APIRef}}

The **`window.requestAnimationFrame()`** method tells the
browser that you wish to perform an animation. I requests that the browser call a
user-supplied callback function prior to the next repaint.

> The frequency of calls to the callback function will generally match the display
> refresh rate. The most common refresh rate is 60hz,
> (60 cycles/frames per second), though 75hz, 120hz and 144hz are also widely used.

Reviewer: ```suggestion
The frequency of calls to the callback function will generally match the display
refresh rate. The most common refresh rate is 60hz,
(60 cycles/frames per second), though 75hz, 120hz, and 144hz are also widely used.
```

Author: I see you like the Oxford/Harvard/serial comma - I do too, but I figured that MDN used the standard American comma.  Is that defined somewhere in a style guide here?

Reviewer: I’m not sure whether or not it is

> The frequency of calls to the callback function will generally match the display
> refresh rate, per the W3C recommendation. The most common refresh rate is 60hz,
> (60 cycles/frames per second), though 75hz, 120hz and 144hz are also widely used.

Reviewer: ```suggestion
The frequency of calls to the callback function will generally match the display
refresh rate. The most common refresh rate is 60hz,
(60 cycles/frames per second), though 75hz, 120hz and 144hz are also widely used.
```

`requestAnimationFrame()` calls are paused in most browsers when running in

> background tabs or hidden {{ HTMLElement("iframe") }}s in order to improve
> performance and battery life.
> 
> > **Note:** Your callback function must call `requestAnimationFrame()` again if
> > you want to animate another frame. `requestAnimationFrame()` is one-shot.
> 
> > **Warning:** Be sure to always use the first argument (or some other method for
> > getting the current time) to calculate how much the animation will progress in
> > a frame, **otherwise the animation will run faster on high refresh rate screens**.
> > See the examples belows for ways to do this.

Reviewer: ```suggestion
background tabs or hidden {{ HTMLElement("iframe") }}s, in order to improve
performance and battery life.

> **Note:** Your callback function must call `requestAnimationFrame()` again if
> you want to animate another frame. `requestAnimationFrame()` is one-shot.

> **Warning:** Be sure to always use the first argument (or some other method for
> getting the current time) to calculate how much the animation will progress in
> a frame — **otherwise the animation will run faster on high refresh-rate screens**.
> For ways to do that, see the examples below.
```

### Parameters

- `callback`
  - The function to call when it's time to update your animation for the next
    repaint. This callback function is passed a single argument: a
    {{domxref("DOMHighResTimeStamp")}} indicating the end time of the previous frame's
    rendering (based on the number of milliseconds since
    [time origin](/en-US/docs/Web/API/DOMHighResTimeStamp#the_time_origin)).

>   - The timestamp is a decimal number, in milliseconds, but with a minimal
>     precision of 1 millisend. For `Window` objects (not `Workers`), it is equal to
>     {{domxref("document.timeline.currentTime")}}. This timestamp is shared

Reviewer: ```suggestion
  - The timestamp is a decimal number, in milliseconds, but with a minimal
    precision of 1 millisecond. For `Window` objects (not `Workers`), it is equal to
    {{domxref("document.timeline.currentTime")}}. This timestamp is shared
```

>   - The function to call when it's time to update your animation for the next
>     repaint. This callback function is passed a single argument, a
>     {{domxref("DOMHighResTimeStamp")}} indicating end time of the previous frame's
>     rendering (based on the number of milliseconds since
>     [time origin](/en-US/docs/Web/API/DOMHighResTimeStamp#the_time_origin)).
>   - The timestamp is a decimal number, in milliseconds, but with a minimal
>     precision of 1ms. For `Window` objects (not `Workers`), it is equal to
>     {{domxref("document.timeline.currentTime")}}. This timestamp is shared
>     between all windows that run on the same agent (i.e. all same origin windows,
>     and more importantly same-origin iframes), which allows synchronizing
>     animations across multiple `requestAnimationFrame` callbacks. The timestamp
>     value is also similar to calling {{domxref('performance.now()')}} at the start
>     of the callback function, but it is never the same value.
>   - When multiple callbacks queued by `requestAnimationFrame()` begin to fire in
>     a single frame, each receives the same timestamp even though time has passed
>     during the computation of every previous callback's workload.

Reviewer: ```suggestion
  - The function to call when it's time to update your animation for the next
    repaint. This callback function is passed a single argument: a
    {{domxref("DOMHighResTimeStamp")}} indicating the end time of the previous frame's
    rendering (based on the number of milliseconds since
    [time origin](/en-US/docs/Web/API/DOMHighResTimeStamp#the_time_origin)).
  - The timestamp is a decimal number, in milliseconds, but with a minimal
    precision of 1 millisend. For `Window` objects (not `Workers`), it is equal to
    {{domxref("document.timeline.currentTime")}}. This timestamp is shared
    between all windows that run on the same agent (that is, all same-origin windows,
    and more importantly same-origin iframes) — which allows synchronizing
    animations across multiple `requestAnimationFrame` callbacks. The timestamp
    value is also similar to calling {{domxref('performance.now()')}} at the start
    of the callback function, but it is never the same value.
  - When multiple callbacks queued by `requestAnimationFrame()` begin to fire in
    a single frame, each receives the same timestamp even though time has passed
    during the computation of every previous callback's workload.
```

Author: There's a typo here: "1 millisend" should be "1 millisecond".

- `callback`
  - : The function to call when it's time to update your animation for the next repaint.
    The callback function is passed one single argument, a {{domxref("DOMHighResTimeStamp")}}
    indicating end time of the previous frame's rendering.
    For `Window` objects (not `Workers`), the timestamp's value is equal to
    {{domxref("document.timeline.currentTime")}}. This value is shared between
    all windows that run on the same agent (i.e. all same origin windows, and more
    importantly same-origin iframes), which allows synchronizing an animation across
    multiple `requestAnimationFrame` callbacks. The timestamp value is also similar to

>     {{domxref('performance.now()')}}, but never the same value.

Reviewer 2: perhaps similar to calling performance.now() at the beginning of the callback?

Author: Based on our discussions here I was shying away from performance.now() here, but I can certainly add something to that effect.  I'll take your comment as a vote to keep the performance.now example too.

Reviewer 2: I meant, instead of saying "similar to `performance.now()`" say "similar to calling `performance.now()` at the beginning of the callback" or so.

Reviewer 2: Great. It was a nit, I'll try to find time for the other stuff. Thanks for working on this!

- `callback`
  - : The function to call when it's time to update your animation for the next repaint.
    The callback function is passed one single argument, a
    {{domxref("DOMHighResTimeStamp")}} similar to the one returned by
    {{domxref('performance.now()')}}, indicating the point in time when
    `requestAnimationFrame()` starts to execute callback functions. The differences
    between this argument and {{domxref('performance.now()')}} are worth noting. First,
    in order to deal with a security issue, all the browsers reduced the precision of
    {{domxref('performance.now()')}} to varying degrees, and Firefox returns an integer.
    This argument also has its precision limited as illustrated by the limited number of
    decimal places it uses, again varying by browser. In terms of the values themselves,
    the {{domxref('performance.now()')}} value is always higher. The amount higher varies

>     by browser and momentary circumstances, but in some cases it can be almost as much as

Reviewer 3: It can actually be an arbitrarily large amount of time, e.g. in cases where a previous callback gets executed in the same frame and takes arbitrarily long time to execute.

```js
requestAnimationFrame((t1) => {
	while(performance.now() - t1 < 2000) {}
})
requestAnimationFrame((t1) => {
	console.log("late by", performance.now() - t1); // 2000.xx
})
```

Author: Yes it can.  I've also seen it happen with GPU overload when animating complex filters, outside of any javascript delay.  I will amend this to reflect your comment.  I hope to have some time over the next few days to get to it.  Based on other comments and reading that I did after filing this PR I'll be taking more of a chainsaw to this than a surgical knife, so I won't be making a tidy edit here in review-land, but I'll post a comment along with my push.  Thanks.

Author: @Kaiido - I misunderstood this comment last night, and though it is true and the code does log  differences > 2 seconds, this seems like an unusual setup, calling `requestAnimationFrame()` twice like that outside of a callback loop.
In practice, when would someone write code like this without looping in the callback?  This code (in [this codepen](https://codepen.io/sidewayss/pen/BavNmMq)) is the same concept but in a callback loop, and the timestamp is never much more than a frame behind `performance.now()`:
```
function raf(t) {
  const now = performance.now();
  console.log(t, now , t - now);
  while(performance.now() - t < 1000) {};
  if (x--)
    requestAnimationFrame(raf);  
}
let x = 3;
requestAnimationFrame(raf);
```

Reviewer 3: Yes, having the same script require multiple times an animation frame is unusual and even a sign of bad code, however it's rather common to have multiple scripts / components that will require their own animation frame for various reasons, e.g. some scrolling effect + some continuous animation + some  event debouncing + etc. So it's actually pretty common to have `requestAnimationFrame` callbacks that end up being in concurrency with other callbacks.

Author: @Kaiido - Does the mousemove event throttling affect all operations that involve moving the mouse, such as the Window resize event? I have throttled the resize event in the past and I'm wondering if I didn't need to do so.

I ran your fiddle on Chrome and Firefox and mostly see what you describe.  What is "double frame buffering"?  Also, unfortunately, the results I see in Firefox are more erratic than what you describe.  Sometimes the direct call does not have the 2s delay time, and sometimes the timeout call has the 2s delay.

Author: Thanks for the explanations.  The Firefox discrepancies appeared several times as I fiddled, but they seem to appear rarely.  I'm on Windows, fwiw.
